contracts/RewardController.sol

```diff
@@ -0,0 +1,252 @@
+        uint256 allocPoint;
+    }
+
+    struct VaultUpdate {
+        uint256 blockNumber;// update blocknumber
+        uint256 totalAllocPoint; //totalAllocPoint
+    }
+
+    uint256 public constant MULTIPLIERS_LENGTH = 24;
+
+    // Block from which the vaults contract will start rewarding.
+    uint256 public startBlock;
+    uint256 public epochLength;
+    // the ERC20 contract in which rewards are distributed
+    IERC20Upgradeable public rewardToken;
+    // Reward Multipliers
+    uint256[24] public rewardPerEpoch;
+    VaultUpdate[] public globalVaultsUpdates;
+    mapping(address => VaultInfo) public vaultInfo;
+    // vault address => user address => reward debt amount
+    mapping(address => mapping(address => uint256)) public rewardDebt;
+    // vault address => user address => unclaimed reward amount
+    mapping(address => mapping(address => uint256)) public unclaimedReward;
+
+    event SetRewardPerEpoch(uint256[24] _rewardPerEpoch);
+    event SafeTransferReward(
+        address indexed user,
+        address indexed vault,
+        uint256 amount,
+        address rewardToken
+    );
+    event ClaimReward(address indexed _vault, address indexed _user, uint256 _amount);
+
+    function initialize(
+        address _rewardToken,
+        address _hatGovernance,
+        uint256 _startRewardingBlock,
+        uint256 _epochLength,
+        uint256[24] memory _rewardPerEpoch
+    ) external initializer {
+        rewardToken = IERC20Upgradeable(_rewardToken);
+        startBlock = _startRewardingBlock;
+        epochLength = _epochLength;
+        rewardPerEpoch = _rewardPerEpoch;
+        _transferOwnership(_hatGovernance);
+    }
+
+    function setAllocPoint(address _vault, uint256 _allocPoint) external onlyOwner {        
+        updateVault(_vault);
+        uint256 totalAllocPoint = (globalVaultsUpdates.length == 0) ? _allocPoint :
+        globalVaultsUpdates[globalVaultsUpdates.length-1].totalAllocPoint - vaultInfo[_vault].allocPoint + _allocPoint;
+        if (globalVaultsUpdates.length > 0 &&
+            globalVaultsUpdates[globalVaultsUpdates.length-1].blockNumber == block.number) {
+            // already update in this block
+            globalVaultsUpdates[globalVaultsUpdates.length-1].totalAllocPoint = totalAllocPoint;
+        } else {
+            globalVaultsUpdates.push(VaultUpdate({
+                blockNumber: block.number,
+                totalAllocPoint: totalAllocPoint
+            }));
+        }
+
+        vaultInfo[_vault].allocPoint = _allocPoint;
+    }
+
+    /**
+    * @notice Safe HAT transfer function, transfer rewards from the contract only if there are enough
+    * rewards available.
+    * @param _to The address to transfer the reward to
+    * @param _amount The amount of rewards to transfer
+    * @param _vault The vault address
+   */
+    function safeTransferReward(address _to, uint256 _amount, address _vault) internal {
+        if (rewardToken.balanceOf(address(this)) < _amount)
```

Comment:
perhaps we can remove this check, as the transfer call will revert anyway

Reply:
actually, let's remove this function completely, it is only used in claimReward. And that has its own event. So we can also remove teh SafeTransferREward event.